contracts: Make L2 genesis script configurable

sebastianst opened the pull request:
<!-- Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md -->

**Description**

Makes the L2 genesis script configurable via two new env vars
* `OUTPUT_MODE`
    * `none` - no state dump file written, used in tests
    * `latest` - only dump state of latest fork
    * `all` - dump states of all forks (default)
* `FORK` - for fork selection (`delta`, `ecotone`,`fjord` or `latest` (default, always selects latest fork))

All generated devnet allocs now explicitly have their fork as suffix, also the latest (no suffix previously).

Still WIP:
- [ ] solidity documentation needs update

**Additional context**

The L2 genesis script before had the latest fork hardcoded.

**Metadata**

- Fixes https://github.com/ethereum-optimism/client-pod/issues/895

tynes commented:
These abstractions definitely look clean to me! I am wondering since because the deploy config has access to the fork offsets, we could simplify things and remove the need to configure the fork via env var and instead parse the forks that have a genesis offset of 0

I imagine this looking like in the deploy config:

```solidity
fjordOffsetTimestamp = _envOr(json, "fjordOffsetTimestamp", type(uint256).max);
```

then you could have something like `DeployConfig.latestFork()` or something like that or `isXYZ`

Just suggesting this to reduce possible footguns. Besides that, this code is clean and well structured

sebastianst committed:
contracts: Make L2 genesis script configurable

sebastianst commented:
@tynes thanks for the suggestion, implemented (hopefully something like) this in the 2nd commit. If the `FORK` env var is set, it takes precedent, otherwise the fork is read from the deploy config. If the deploy config isn't set either, it fails to run, which is probably good because if forces the user to actively think about which fork to set.

If the 2nd commit isn't going in the right direction or needs much more work, we can also split it out into a 2nd PR and get the 1st commit in so that the functionality for fork selection is already there. So chain providers asking for this already have something that works.

sebastianst committed:
contracts: Read latest fork from deploy config in L2 genesis script

sebastianst commented:
I should probably add some documentation about the new env vars.

sebastianst committed:
contracts: add README entry for new L2 genesis script env vars

tynes commented on packages/contracts-bedrock/scripts/L2Genesis.s.sol:
```diff
@@ -120,7 +106,21 @@ contract L2Genesis is Deployer {
+    }
+
+    function latestDeployConfigGenesisFork() internal returns (Fork) {
+        DeployConfig cfg = DeployConfig(address(uint160(uint256(keccak256(abi.encode("optimism.deployconfig"))))));
```
Is there a reason as to why we cannot use the globally defined deploy config?

sebastianst replied:
I totally should use that globally defined deploy config 😅 I wasn't aware it's there as I'm a total forge script noob 😬

tynes replied:
All good!

tynes commented on packages/contracts-bedrock/scripts/L2Genesis.s.sol:
```diff
@@ -120,7 +106,21 @@ contract L2Genesis is Deployer {
+            fork = latestDeployConfigGenesisFork();
+            console.log("L2Genesis: using deploy config fork: %d", uint256(fork));
+        } else {
+            console.log("L2Genesis: using env var fork: %d", uint256(fork));
```
It would be way more useful if there was a function that converts the `uint256` to a `string` for these logging statements

sebastianst replied:
It would, I just followed the pattern I saw elsewhere where we, instead of implementing string conversions, we just log the enum number representation. But I can add that string converter.

tynes replied:
Its true that overall we could be better about this 😅 at least in this case, it would be good for chain deployers having insight into the config that is being used

tynes replied:
We would be able to get rid of `latestDeployConfigGenesisFork` if we have `Config.fork` be the source of truth for both config file + env var

tynes commented on packages/contracts-bedrock/scripts/L2Genesis.s.sol:
```diff
@@ -147,28 +154,29 @@ contract L2Genesis is Deployer {
-            return;
-        }
-        if (_mode == OutputMode.OUTPUT_ALL) {
+        if (_mode == OutputMode.ALL || _fork == Fork.DELTA && _mode == OutputMode.LATEST) {
```
Could we make the logic inside of the `if` expression an internal function that accepts a `Fork` and returns a `bool`? Then its really simple to add new forks

sebastianst replied:
done, encapsulated in `writeForkGenesisAllocs`

tynes commented on packages/contracts-bedrock/scripts/Config.sol:
```diff
@@ -67,4 +87,44 @@ library Config {
+    ///         unset, NONE is returned.
+    ///         If set to the special value "latest", the latest fork is returned.
+    function fork() internal view returns (Fork) {
+        string memory forkStr = vm.envOr("FORK", string(""));
```
I do think we can simplify by removing `Fork.NONE`. I also think it would be better to call it `Fork.BEDROCK`

tynes replied:
Thoughts on removing `Fork.NONE` for now and then we can reintroduce it in the future as `Fork.BEDROCK` when we standardize on the predeploy upgrade flow?

sebastianst replied:
I also wasn't super happy about adding the `Fork.NONE` option, but I need it to signal back from the `Config.fork()` function that the env var is not set, so I can then fallback to the deploy config one in the L2 genesis script.

tynes replied:
Ah ok, this makes sense, thoughts on just renaming it to `BEDROCK` for now?

tynes replied:
The default could be `BEDROCK` and we could remove the length check?

sebastianst replied:
Btw note that `NONE` is not Bedrock, but stands rather for _unspecified_.

sebastianst replied:
I need a way to signal from `Config` that the env var is not set. I didn't intend to support Bedrock, Regolith or Canyon, as we didn't before and I guess nobody needs it really.

tynes commented on packages/contracts-bedrock/scripts/DeployConfig.s.sol:
```diff
@@ -161,6 +175,17 @@ contract DeployConfig is Script {
         useInterop = _readOr(_json, "$.useInterop", false);
     }
 
+    function latestGenesisFork() public view returns (Fork) {
```
Thoughts on using `Config.Fork` here and update this function to be `fork()` and then prioritize the deploy config, if not present fall back to the env var?

sebastianst replied:
I thought the `DeployConfig` should be the sole abstraction around the deploy config json file, whereas the `Config` is for reading env vars etc. So as to not mix abstractions I preferred to handle the fallback logic in the L2 genesis script. But I can see that it's nicer to encapsulate the config loading logic into one of these instead of at the consumer end, in the L2 genesis script.

Wondering if we want to make one or the other the higher abstraction. E.g. we could make `Config` the higher abstraction that accesses `DeployConfig` in its `Config.fork()` function. For this I'd just need to move the two L2 genesis enums `OutputMode` and `Fork` into their own solidity file, like `scripts/GenesisConfig.s.sol`.

tynes replied:
I definitely see your thinking regarding keeping the abstractions thin, from a consumer pov I do think its nice to have a single place to read config relevant to deployment, which to me is the `DeployConfig`. We don't really have this pattern of env vars in the `DeployConfig` today, but i do think that we want a single source of truth where the fork is pulled from which I think that should be the `DeployConfig` and we can start wrapping `Config` with the `DeployConfig` for the deploy config -> env var fallback, so that devs can always just use `DeployConfig` for the source of ultimate truth, enabling a consistent way that files and env vars relate to each other with the deploy config, ie which takes precedence

sebastianst replied:
That also makes sense, added `fork()` to `DeployConfig`, which internally still calls `Config.fork()` and if this is unset, reads the latest genesis fork from the deploy config.

tynes commented:
I think there are some ways to simplify this, left comments, curious what you think @sebastianst

tynes commented on packages/contracts-bedrock/scripts/DeployConfig.s.sol:
```diff
@@ -161,6 +176,30 @@ contract DeployConfig is Script {
+        }
+    }
+
+    function latestGenesisFork() public view returns (Fork) {
```
Thoughts on having just 1 public function vs 2?

sebastianst replied:
Even though `latestGenesisFork` is not used publicly right now, I left it public because it conveys a different information about the deploy config that could be used in the future by another contract. I thought it doesn't really matter. I will make it internal and if we ever need it as a public fn, can make it public again.

tynes commented:
Looks like linting has failed

tynes commented:
If you rebase on `develop`, then it will fix `ci/circleci: sdk-tests`

tynes commented on packages/contracts-bedrock/test/setup/Setup.sol:
```diff
@@ -175,9 +177,10 @@ contract Setup {
     /// @dev Sets up the L2 contracts. Depends on `L1()` being called first.
     function L2() public {
-        console.log("Setup: creating L2 genesis, with output mode %d", uint256(l2OutputMode));
+        console.log("Setup: creating L2 genesis with fork %d", uint256(l2Fork));
```
Should this be `l2Fork.toString()`?

sebastianst replied:
yes, overlooked when switching to `toString()` 🙏🏻

sebastianst committed:
address review
